---
title: "Adv. Immo Tutorial"
author: "Bernd Jagla"
date: "11/22/2019"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Getting started

Load the app using RStudio and calling 

```
library(ideal)
ideal()
```

This will open a web-browser with the application. Please be a bit patient as it sometimes takes a few seconds to load everything. This holds true for most of the tabs. The only way to check if there is still something running is to check the process manager.


```{r, out.width='90%', fig.align='center', fig.cap='Overview, showing entry screen', echo=FALSE}
knitr::include_graphics('figures/tut1.png')
```

After the introduction screen is opened we are ready to load the data. To do so, please click on "Data Setup" (see arrow in figure above).


```{r, out.width='90%', fig.align='center', fig.cap='Data loading 1', echo=FALSE}
knitr::include_graphics('figures/tut2.png')
```

Here, we need to provide the actual count data (1) and the meta-data or information about the samples (2).

1. The count file is a comma-separated text file where the first row holds the sample names and the first column holds the gene names.
We use a manually modified version where we have changed the column names to follow the following convention. The name "D1a_O_L_L" can be read as **D**onor **1** of year **a** (a=1st year, b=2nd year,...) **_** who is **old** (vs. Y = young) **_** and was CMV positive (**P_**) (vs N) and was stimulated with LPS **L** (L = LPS, A = antiCD3/CD8, n = no stimmulation, P = polyIC). This conversion is done with a simple text editor. Excel could be used for this, but this has the tendancy to change values (e.g. Sep07 will be interpreted as a data and consequently be changed). Please see (<a style="color:red">REF</a>) for a more detailed discussion. 

We are providing the data from the previous years in the file called "allPreviousYears.csv" and the results from this year in the files "<a style="color:red">CHANGEME.xls</a>" und <a style="color:red">CHANGEME.xls</a>"

```{r, out.width='90%', fig.align='center', fig.cap='A CSV file with the count data', echo=FALSE}
knitr::include_graphics('figures/tut3.png')
```

Though we have already encode most of the information about the samples in the name we still need to tell the program how to interprete the data. This, we do again using a TSV file that looks like this:

```{r, out.width='90%', fig.align='center', fig.cap='A tab delimited file with the meta data', echo=FALSE}
knitr::include_graphics('figures/tut4.png')
```

In this case, we are using "tabulators" instead of commas to separate the columns. Please also note the lines starting with **#**. Those lines will be ignored from both the meta data and the count data files. We have already removed most of the samples that didn't produce any useful data. There are still some that might be useable when sufficient other material is collected this year.

It is up to your biological and scientific knowledge/experience and scientific question that will guide you to what to include and exclude.

We can load multiple files. (This is one multiple modificatios I have introduced for this version of IDEAL.)
We will be loading the data from all previous years and this year (2019), see following figure.

```{r, out.width='90%', fig.align='center', fig.cap='Load multiple count data files', echo=FALSE}
knitr::include_graphics('figures/tut5.png')
```

Next, we load the meta data (2) using the file "target.all.txt". (No figure)

Please note, that as a result we notice changes highlighted in the following figure:


```{r, out.width='90%', fig.align='center', fig.cap='Step 2', echo=FALSE}
knitr::include_graphics('figures/tut6.png')
```


## Step 2

Now, we have to choose the design of our experiment under "Step 2".

<a style="color:red">Excurse: Design/design matrix.</a>


Based on what we learned about the design of an experiment we are going to set the model to 

Stimulus + CMVstatus + age, with Stimulus = null as the reference.

Ideally, we would include the donor as a <a style="color:red">WHAT???</a>, but we don't have enough replicats.

```{r, out.width='90%', fig.align='center', fig.cap='Step 2 filled', echo=FALSE}
knitr::include_graphics('figures/tut7.png')
```

After we generated the DESeq2 object, we can verify that all the parameters are correct:

```{r, out.width='90%', fig.align='center', fig.cap='verify DESeq2 object', echo=FALSE}
knitr::include_graphics('figures/tut8.png')
```

We can now also load additional annotation and select samples we would like to exclude (potential outliers). The annotation is needed later when we want to perform gene set enrichment analysis and other data integration processes. We will also see later during the quality control and data anlaysis which samples to exclude. The parameters for the annotation are already preset correctly. When working with different sources of count data this might have to be adjusted. This is dependant on the reference genome used for the alignment and the type of identifiers used. This is briefly described during the general bioinformatics presentation.

```{r, out.width='90%', fig.align='center', fig.cap='Annotation and sample selection', echo=FALSE}
knitr::include_graphics('figures/tut9.png')
```

We can now calculate all the coefficients for the design / model that we selected. This will take a minute or two.

```{r, out.width='90%', fig.align='center', fig.cap='Run DESeq2', echo=FALSE}
knitr::include_graphics('figures/tut10.png')
```

After this has finished we are ready to look at some of the results. We first see how many genes are up/down regulated

```{r, out.width='90%', fig.align='center', fig.cap='DESeq2 finished', echo=FALSE}
knitr::include_graphics('figures/tut11.png')
```


```{r, out.width='90%', fig.align='center', fig.cap='Dispersion plot', echo=FALSE}
knitr::include_graphics('figures/tut12.png')
```

We can equally see how the samples compare when looking at the genes with no expression (zero counts), and the number of aligned sequences.
The two table are there to check that the input has been correctly loaded.

```{r, out.width='90%', fig.align='center', fig.cap='Summary stats', echo=FALSE}
knitr::include_graphics('figures/tut13.png')
```

In the "Counts overview" we can perform mnore quality controls.

```{r, out.width='90%', fig.align='center', fig.cap='Count overview', echo=FALSE}
knitr::include_graphics('figures/tut14.png')
```

1). The checkmarks indicate what we have done so far.
2). Similar information is available on the top of the page
3). We have now moved to the "Counts overview" tab.

The following tables are available

```{r, out.width='90%', fig.align='center', fig.cap='Selection for table', echo=FALSE}
knitr::include_graphics('figures/tut15.png')
```


<a style="color:red">Explain different tables</a>


Filter out genes that are not well expressed.

```{r, out.width='90%', fig.align='center', fig.cap='Filter genes', echo=FALSE}
knitr::include_graphics('figures/tut16.png')
```


Pearson
Spearman
Kendal 
Distance measures

Pairwise correlation


```{r, out.width='90%', fig.align='center', fig.cap='Pairwise correlation', echo=FALSE}
knitr::include_graphics('figures/tut17.png')
```

Correlation between samples shows which samples are correlated using a heatmap. Potentially mislabeled samples can be identified.

```{r, out.width='90%', fig.align='center', fig.cap='Correlation between samples', echo=FALSE}
knitr::include_graphics('figures/tut18.png')
```

We have to set the "Group/color by:" to show the PCA and size factors:


```{r, out.width='90%', fig.align='center', fig.cap='Highest expressing genes', echo=FALSE}
knitr::include_graphics('figures/tut19.png')
```

The PCA and size factor visualizations show us which factors are determining the differences between the samples. I.e. by changing the "group" we can see which PCA groups are built.

We can also identify the highest expressing genes and see who they correlate with the samples.



```{r, out.width='90%', fig.align='center', fig.cap='Extract results', echo=FALSE}
knitr::include_graphics('figures/tut21.png')
```


```{r, out.width='90%', fig.align='center', fig.cap='choose factor', echo=FALSE}
knitr::include_graphics('figures/tut20.png')
```

<a style="color:red">What is the denominator for</a>
<a style="color:red">explain independant filtering</a>
<a style="color:red">Explain IHW</a>

After clicking on the "Extract results" button we have everything for the following analyses.

```{r, out.width='90%', fig.align='center', fig.cap='results are available', echo=FALSE}
knitr::include_graphics('figures/tut22.png')
```

We see that there are 577 genes up-regulated for the comparison null vs. polyIC.

Sortable table with p-values and links to ENSEMBL:

```{r, out.width='90%', fig.align='center', fig.cap='results table', echo=FALSE}
knitr::include_graphics('figures/tut23.png')
```

Diagnostic plots:

P-value distribution

P-value distribution for ranges of mean expression values.

log2 fold change histogram.


```{r, out.width='90%', fig.align='center', fig.cap='results histograms', echo=FALSE}
knitr::include_graphics('figures/tut24.png')
```

Summary Plots

this takes a few seconds to calculate

<a style="color:red">Explain MA plot</a>


Select a region in the MA plot

```{r, out.width='90%', fig.align='center', fig.cap='Summary Plots', echo=FALSE}
knitr::include_graphics('figures/tut25.png')
```

Then click on a gene in right plot.

We then get detailed information on that gene, its expression over the groups that are selected on the left panel and information from the Web on that gene.

```{r, out.width='90%', fig.align='center', fig.cap='info on Genes', echo=FALSE}
knitr::include_graphics('figures/tut26.png')
```



Volcano plot

```{r, out.width='90%', fig.align='center', fig.cap='Volcano plot', echo=FALSE}
knitr::include_graphics('figures/tut27.png')
```

Resizable heatmap of selected genes.

The right heatmap is zoomable.

```{r, out.width='90%', fig.align='center', fig.cap='Heatmap of selected genes', echo=FALSE}
knitr::include_graphics('figures/tut28.png')
```


```{r, out.width='90%', fig.align='center', fig.cap='zoomed heatmap', echo=FALSE}
knitr::include_graphics('figures/tut29.png')
```

Table of zoomed region in MA plot.

```{r, out.width='90%', fig.align='center', fig.cap='Table of zoomed region', echo=FALSE}
knitr::include_graphics('figures/tut30.png')
```

# Gene Finder

Select genes of interest on the left panel.

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut31.png')
```

With additional group

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut32.png')
```

We can also visualize the coefficients for the first gene selected.

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut33.png')
```

Those genes are also highlighted in the MA plot and its statistic values are displayed in a table.

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut34.png')
```

Given a text file with gene names (one per line) we can visualize those in the fashion:
Here we have a list of genes that are associated to cytokine responds.

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut35.png')
```

Functional analysis

```{r, out.width='90%', fig.align='center', fig.cap='Gene Finder', echo=FALSE}
knitr::include_graphics('figures/tut36.png')
```

Gene set enrichment analysis

1. We are now in the Functoinal analysis panel.
2. Select a Gene ontology
3. Select up or down or both or one or two manually defined list of genes.
4. calculate the enrichment
5. the progress is shown on the bottom right.

These calculations will take a few minutes.

<a style="color:red">Explain different GO functions</a>

Table with GO enrichments. Links out to external web-pages.


```{r, out.width='90%', fig.align='center', fig.cap='GO table (goanna)', echo=FALSE}
knitr::include_graphics('figures/tut37.png')
```


GO top table with heatmap.

```{r, out.width='90%', fig.align='center', fig.cap='GO table (gotop)', echo=FALSE}
knitr::include_graphics('figures/tut38.png')
```


Set operations. Here we have loaded again the cytokineResponds.list.txt


```{r, out.width='90%', fig.align='center', fig.cap='set operations', echo=FALSE}
knitr::include_graphics('figures/tut39.png')
```

Once expanded, the table allows identifying the genes that are in each category.

```{r, out.width='90%', fig.align='center', fig.cap='set operations table', echo=FALSE}
knitr::include_graphics('figures/tut44.png')
```


The signature analysis allows to compare to analysis stored for key experiments for signature genes.
Those lists can be retrieved from the Broad institute (http://software.broadinstitute.org/gsea/msigdb/index.jsp).
The immunology gene sets (C7) are the most interesting to us.


```{r, out.width='90%', fig.align='center', fig.cap='signature analysis', echo=FALSE}
knitr::include_graphics('figures/tut40.png')
```

1. we are now in the Signature analysis
2. select the C7 set
3. to make the correspondance between our annotation and the one used by the Broad we set the coresponding values.
4. Here, we compare to the Flu vs. coli infection of PBMC up regulated genes
5. We should cluster by column/sample to have the best visual impression.
6. <a style="color:red">Don't know what this does.</a>

Using per row standardization and only our differentially expressed genes give the following heatmap:

```{r, out.width='90%', fig.align='center', fig.cap='signature analysis heatmap', echo=FALSE}
knitr::include_graphics('figures/tut41.png')
```


## Report

Creating the preview will take a few minutes.

This allows creating a report of what we have done for reproducibility.

```{r, out.width='90%', fig.align='center', fig.cap='Report generator', echo=FALSE}
knitr::include_graphics('figures/tut42.png')
```

There are several option we can choose from to create a report, either in R-markdown or as an HTML report

```{r, out.width='90%', fig.align='center', fig.cap='Report generator with options', echo=FALSE}
knitr::include_graphics('figures/tut43.png')
```














